draft-irtf-cfrg-vdaf.md

```diff
@@ -1190,6 +1190,175 @@ how the nonces are chosen, but security requires that the nonces be unique. See
+  computes its second-round prep share locally, computes its output share, and
+  sends the second-round prep message to the Helper. Finally, the Helper
+  computes its own output share. Preparation requires two requests.
+
+* In general, each request includes the Leader's prep share for the previous
+  round and/or the prep message for the current round; correspondingly, each
+  response consists of the prep message for the current round and the Helper's
+  prep share for the next round.
+
+The Aggregators proceed in this ping-ponging fashion until a step of the
+computation fails, indicating the report is invalid and should be rejected, or
+the last round is reached and each Aggregator computes an output share. All
+told there there are `ceil((Vdaf.ROUNDS+1)/2)` requests sent.
+
+~~~ transition
+def ping_pong_next(Vdaf,
+                   state: State,
+                   is_leader: bool,
+                   agg_param: Vdaf.AggParam,
+                   inbound: Optional[Message],
+                   ) -> (State, Optional[Message]):
+    if type(inbound) == InitializeMessage:
+        prep_msg, peer_prep_share = None, inbound.prep_share
+    elif type(inbound) == ContinueMessage:
+        prep_msg, prep_share = inbound.prep_msg, inbound.prep_share
+    elif type(inbound) == FinishMessage:
+        prep_msg, peer_prep_share = inbound.prep_msg, None
+    else:
+        prep_msg, peer_prep_share = None, None
+
+    try:
+        # If prep_msg == None then this is the start of the
+        # first round. Otherwise, this is the prep message
+        # from the end of the previous round.
+        out = Vdaf.prep_next(host_prep_state, prep_msg)
+        if type(out) == Vdaf.OutShare:
+            # The Leader reaches this state upon processing a
+            # response from the Helper, so there is no outbound
+            # message.
+            outbound = None if is_leader else FinishedMessage()
+            return (Finished(out), outbound)
+        (host_prep_state, host_prep_share) = out
+
+        if peer_prep_share != None:
+            prep_shares = [peer_prep_share, host_prep_share]
+            if is_leader:
+                prep_shares.reverse()
+            prep_msg = Vdaf.prep_shares_to_prep(
+                agg_param,
+                prep_shares,
+            )
+            out = Vdaf.prep_next(host_prep_state, prep_msg)
+            if type(out) == Vdaf.OutShare:
+                return (Finished(out), FinishMessage(prep_msg))
+            (host_prep_state, host_prep_share) = out
+
+        if is_leader and inbound == None:
+            outbound = InitializeMessage(host_prep_share)
+        else:
+            outbound = ContinueMessage(prep_msg, host_prep_share)
+        return (Continued(host_prep_state), outbound)
+    except:
+        return (Rejected(), None)
+~~~
+
+At the start of the first request/response cycle, the Leader would call
+
+~~~
+outbound = Vdaf.ping_pong_next(True, state, agg_param, None)
```

Comment:
nit: `state` comes before `is_leader`, here and on L1341.

Reply:
Good catch.